### ISSUE TITLE
Execute action email link - direct to the action itself

### ISSUE THREAD

**Issue by moriyasror** [keycloak/keycloak#25719]
### Description

Today when we call the execute-action-email with redirect:
1. user get email with link
2. link open page with the text: "Perform the following action(s.." and link to perform the action
3. after action was done there is another page saying "Your account has been updated." and link to return to application (if we send redirect uri)
![image](https://github.com/keycloak/keycloak/assets/16653862/d49f2304-ef02-4a38-ac4e-e2f0b9192059)

We want that 
1. step 2, the link in the email will open the action itself without "middle paging" with link to the action. (the same as in forgot password flow)
2. if there is a redirect uri to return to it directly without the middle page and the link "back to application" 

### Motivation

those middle pages are bad UX, why user need to click twice in order to update password

**Comment by ricotero10** [comment]
In my org. we find the same issue. It affects negatively the user experience
Please allow providing direct links both to the system and in redirections

**Comment by AlexShimanovich** [comment]
Same here.
This screen and links redundancy is not up to the UX/UI standards we maintain in our company.
As we all know, no matter how brilliant the product is, the users main concern is always the screens and buttons of the UI.

**Comment by jonkoops** [comment]
Mhhh, yeah this looks like valid feedback to me, we can do better here. @xianli123 WDYT?

**Comment by xianli123** [comment]
@jonkoops  I agree with you. In step 2 @moriyasror shared, we can remove the "Click here to proceed" link and directly show the actions. I attached an example below as shown in the first snapshot.  
<img width="583" alt="Screenshot 2024-01-04 at 14 17 19" src="https://github.com/keycloak/keycloak/assets/53931519/630f63a4-5b0f-4182-b56e-5dc52b769540">

If there is a redirect URI along with the action, we can lead users to the redirect URI after they finish the actions. Let's take updating password as an example, users will enter the login page after updating the password as shown below. The alter indicated the actions' result.
<img width="602" alt="Screenshot 2024-01-04 at 14 25 45" src="https://github.com/keycloak/keycloak/assets/53931519/d8651ec8-db7c-49e1-b31b-5dd4057d4b7f">

Does it make sense? @jonkoops @moriyasror @AlexShimanovich @ricotero10 The screens are not so accurate. I made this simple design rarely based on your requirements and suggestions. Please ignore the details and just focus on the whole process.

**Comment by xianli123** [comment]
@moriyasror Could you please share the steps to reproduce the flow? I want to dig into some details. Thanks!

**Comment by jonkoops** [comment]
> if there is a redirect uri we should return to it directly without the middle page and the link "back to application"

Let's handle this as a separate issue to avoid cluttering the discussion. Could you log an issue for this?

**Comment by jonkoops** [comment]
@xianli123 as mentioned above, let's focus only on the form for now and let's not worry about the redirect under this issue. I think your design is fine for what we need to accomplish here, I am only left wondering what happens if the user is requested to perform multiple steps (e.g. update their account information).

**Comment by jonkoops** [comment]
Another thing I find strange is that this flow is different from if you request a password update through the Administration Console. In that case the user just gets prompted directly to update their password: 

![Screen Shot 2024-01-04 at 13 10 56](https://github.com/keycloak/keycloak/assets/695720/981530f4-6d8a-47d0-bd81-217fdfb55a41)

@moriyasror did you test this behaviour in the latest version of Keycloak, or are you running an older version?

**Comment by xianli123** [comment]
@jonkoops I considered the case with multiple steps. How about using a wizard to list the steps?

<img width="814" alt="Screenshot 2024-01-04 at 20 32 01" src="https://github.com/keycloak/keycloak/assets/53931519/6855c06d-4dd4-4241-ad4a-d1470f9b778a">

**Comment by jonkoops** [comment]
@xianli123 Yeah that is an option. Personally, I think it should work the way the existing flows work. For example, if I request a user reset their password and update their profile these will be two separate steps, without any additional information.

I would go the following way:

- If a user has to perform multiple steps, show them an overview of actions that the user needs to perform, with a button to start the process. After this step the list of required actions is no longer visible.
- If the user only has a single step to perform, then show that action directly.

**Comment by xianli123** [comment]
@jonkoops I agree. It is an easy way to do that. One point we can improve for the multiple steps is to change the link button `Click here to proceed` to a primary button. What do you think?

**Comment by moriyasror** [comment]
@jonkoops the "problematic" update password flow is via the execute-action-email which lead to the middle page (also in v23) the update password from the admin client is when you already login and request update password so it just open the form.

@xianli123 the flow make sense, and will fit our needs. You said to ignore the details but just want to be sure that the current password won't appear in update password from execute-action-email flow. I will explain why

In our product we don't have register flow, we add admin user to our customers and they can add more users. Under the hood when someone add user we add it to Keycloak and run the execute-action-email with 'update password' so he will configure is password and will be able to login to the system. (in this flow there is no current password)
This is the reason why this flow if very critical to us, since every new user get "welcome email" with link to update password which request him to click again on another link and when finish on another link for login.

**Comment by xianli123** [comment]
@moriyasror Thanks. That's an interesting use case. I got you.

**Comment by jonkoops** [comment]
> One point we can improve for the multiple steps is to change the link button `Click here to proceed` to a primary button. What do you think?

Yes, agreed. I think that would be a nice visual enhancement.

**Comment by jgayer** [comment]
This really should be improved, the "Perform the following actions" page makes no sense from a UI perspective and seems like a very developer-centric way of implementing a user interface. The user does not care for "Actions" and "Proceding to do actions". 

someone has written a keycloak extension but that seems a little complicated when you haven't touched the keycloak source code... [Custom SPI - Link for user to set their initial password](https://keycloak.discourse.group/t/custom-spi-link-for-user-to-set-their-initial-password/22157)](https://keycloak.discourse.group/t/custom-spi-link-for-user-to-set-their-initial-password/22157/2)

**Comment by mdiin** [comment]
> @jonkoops the "problematic" update password flow is via the execute-action-email which lead to the middle page (also in v23) the update password from the admin client is when you already login and request update password so it just open the form.
> 
> @xianli123 the flow make sense, and will fit our needs. You said to ignore the details but just want to be sure that the current password won't appear in update password from execute-action-email flow. I will explain why
> 
> In our product we don't have register flow, we add admin user to our customers and they can add more users. Under the hood when someone add user we add it to Keycloak and run the execute-action-email with 'update password' so he will configure is password and will be able to login to the system. (in this flow there is no current password) This is the reason why this flow if very critical to us, since every new user get "welcome email" with link to update password which request him to click again on another link and when finish on another link for login.

In my org we are in the exact same situation, so consider this a point more in favour of this suggestion.

**Comment by hoebelix** [comment]
Our use case is as follows: we don't use the built-in user registration of keycloak since our user registration is on invitation only. So we have our own user registration forms and then create the new user with the REST API. After that we call execute-action-email to let the user validate the email address.

For us it would be ideal if both the page "Perform the following action(s)" and the page "Your account has been updated" would be skipped: the action (verify email address) should be done completely invisible to the user in the background and the user should be redirected directly to the redirect uri.

I think our use case is quite common. Without those middle pages our user registration flow would be as smooth as the user expects and from an ux point of view comparable to the build-in user registration flow of keycloak.

**Comment by mdiin** [comment]
A note on this: We decided to use the [PhaseTwo](https://phasetwo.io/) extension to Keycloak, which implements a magic link login. Using this functionality, we now create users in Keycloak with the API, and trigger a magic link e-mail using the extension's API. This skips the "Perform the following action(s)" page.

**Comment by jeffvictor** [comment]
One simple solution to get around the "Back to Application" page is to override info.ftl in the login theme.  Add the following after the first line:

```
<#if pageRedirectUri?has_content>
	<script type="text/javascript">
		window.location.replace("${pageRedirectUri}");
	</script>
<#elseif actionUri?has_content>
	<script type="text/javascript">
		window.location.replace("${actionUri}");
	</script>
</#if>
```
The entire file (keycloak 25.0.2) should then look like:

```
<#import "template.ftl" as layout>
<#if pageRedirectUri?has_content>
	<script type="text/javascript">
		window.location.replace("${pageRedirectUri}");
	</script>
<#elseif actionUri?has_content>
	<script type="text/javascript">
		window.location.replace("${actionUri}");
	</script>
</#if>
<@layout.registrationLayout displayMessage=false; section>
    <#if section = "header">
        <#if messageHeader??>
            ${kcSanitize(msg("${messageHeader}"))?no_esc}
        <#else>
        ${message.summary}
        </#if>
    <#elseif section = "form">
    <div id="kc-info-message">
        <p class="instruction">${message.summary}<#if requiredActions??><#list requiredActions>: <b><#items as reqActionItem>${kcSanitize(msg("requiredAction.${reqActionItem}"))?no_esc}<#sep>, </#items></b></#list><#else></#if></p>
        <#if skipLink??>
        <#else>
            <#if pageRedirectUri?has_content>
                <p><a href="${pageRedirectUri}">${kcSanitize(msg("backToApplication"))?no_esc}</a></p>
            <#elseif actionUri?has_content>
                <p><a href="${actionUri}">${kcSanitize(msg("proceedWithAction"))?no_esc}</a></p>
            <#elseif (client.baseUrl)?has_content>
                <p><a href="${client.baseUrl}">${kcSanitize(msg("backToApplication"))?no_esc}</a></p>
            </#if>
        </#if>
    </div>
    </#if>
</@layout.registrationLayout>
```

**Comment by Jhereck** [comment]
Any update pls ?

**Comment by Neirda24** [comment]
@mdiin : Hi with your workflow using the magic link. When does the user reset / set its password ?

**Comment by mdiin** [comment]
> [@mdiin](https://github.com/mdiin) : Hi with your workflow using the magic link. When does the user reset / set its password ?

Sorry for the late reply, I got side-tracked by work.

The user is created with a required action of "Update password". This causes the "create password" dialog to be shown when the user clicks the magic link, after which they are sent directly to the application.

**Comment by Ketec** [comment]
Same issue. This step is completely irrelevant and just makes the flow ugly.
Just to get users to set their password.

The whole new password part is already nasty - you have to use "update password", which is also invalid (since it's a new account).

**Comment by abhion** [comment]
hey any update on this?

**Comment by nancysangani** [comment]
@stianst I have opened PR #47466 to fix this issue. Please review it when you get a chance. Thanks!

**Comment by mkunz-it** [comment]
@ahus1 One of my customers would also like to skip both pages. Do we all agree that this would improve the user experience? If so, how can we help to incorporate this new feature into one of the upcoming KeyCloak releases?

**Comment by stianst** [comment]
This can not be added I'm afraid both for a security reason, but also due to some email readers will actively try/scan links in emails

**Comment by Ketec** [comment]
@stianst 
How is this any different from a link to the multi action page?
Or the link in forgot-password email?
